### ISSUE TITLE
Additional documentation needed for scico.solver

### ISSUE THREAD

**Issue by bwohlberg** [lanl/scico#123]
Header docstring of `scico.solver` should include notes on:
- reason for existence of the module (i.e. why the need to wrap the underlying `scipy.optimize` functions)
- when to use the functions in this module over the `jax` optimizers

**Comment by lukepfister** [comment]
We want to *include* functionality from `scipy.optimize` because jax implements a very limited subset of the scipy optimizers: only linear CG and BFGS are fully supported, but there is limited, experimental support for [L-BFGS-B](https://github.com/google/jax/pull/6053)

If you accept that we want the functionality from scipy.optimize, then we hit a problem:  scipy.optimize can only work on 1D, real valued, numpy arrays.  So the wrapper gives the following quality of life improvements:

- Let's us use jax arrays (automatically converts to/from NumPy arrays)
- Enables use of complex arrays (splits into real/imaginary parts)
- Enables use of ND arrays (flattens/reshapes inside of the wrapper)
- jits the function evaluation/gradient

We also hide the use of `jax.grad` for the first-order functions.

(all of this is in the docstring, cf https://github.com/lanl/scico/blob/ec3dea364b03b63578e962cbe90f0d25db7a4fb8/scico/solver.py#L136)


There are tradeoffs compared to using the jax optimizer (including experimental L-BFGS-B)
Pros/Cons of scico.solver:
- Pro:  Fast for small-scale problems.  I don't know why, but jax optimizer has a lot of overhead!
- Pro:  Wraps a *lot* of algorithms, not just L-BFGS and BFGS
- Con:  Can't jit solver, can't take gradients through the solver
- Con:  Inefficient IO; lots of host-device copies (DeviceArrays are cast to NumPy arrays before each function evaluation)
